src/main/scala/org/graphframes/examples/BeliefPropagation.scala

```diff
@@ -0,0 +1,259 @@
+    val sc = SparkContext.getOrCreate(conf)
+    val sql = SQLContext.getOrCreate(sc)
+
+    // Create graphical model g of size 3 x 3.
+    val g = gridIsingModel(sql, 3)
+
+    println("Original Ising model:")
+    g.vertices.show()
+    g.edges.show()
+
+    // Run BP for 5 iterations.
+    val numIter = 5
+    val results = runBPwithGraphX(g, numIter)
+
+    // Display beliefs.
+    val beliefs = results.vertices.select("id", "belief")
+    println(s"Done with BP. Final beliefs after $numIter iterations:")
+    beliefs.show()
+
+    sc.stop()
+  }
+
+  /**
+   * Given a GraphFrame, choose colors for each vertex.  No neighboring vertices will share the
+   * same color.  The number of colors is minimized.
+   *
+   * This is written specifically for grid graphs. For non-grid graphs, it should be generalized,
+   * such as by using a greedy coloring scheme.
+   *
+   * @param g  Grid graph generated by [[org.graphframes.examples.Graphs.gridIsingModel()]]
+   * @return  Same graph, but with a new vertex column "color" of type Int (0 or 1)
+   */
+  private def colorGraph(g: GraphFrame): GraphFrame = {
+    val colorUDF = udf { (i: Int, j: Int) => if ((i + j) % 2 == 0) 0 else 1 }
+    val v = g.vertices.withColumn("color", colorUDF(col("i"), col("j")))
+    GraphFrame(v, g.edges)
+  }
+
+  /**
+   * Run Belief Propagation.
+   *
+   * This implementation of BP shows how to use GraphX's aggregateMessages method.
+   * It is simple to convert to and from GraphX format.  This method does the following:
+   *  - Color GraphFrame vertices for BP scheduling.
+   *  - Convert GraphFrame to GraphX format.
+   *  - Run BP using GraphX's aggregateMessages API.
+   *  - Augment the original GraphFrame with the BP results (vertex beliefs).
+   *
+   * @param g  Graphical model created by [[org.graphframes.examples.Graphs.gridIsingModel()]]
+   * @param numIter  Number of iterations of BP to run.  One iteration includes updating each
+   *                 vertex's belief once.
+   * @return  Same graphical model, but with [[GraphFrame.vertices]] augmented with a new column
+   *          "belief" containing P(x,,i,, = +1), the marginal probability of vertex i taking
+   *          value +1 instead of -1.
+   */
+  def runBPwithGraphX(g: GraphFrame, numIter: Int): GraphFrame = {
+    // Choose colors for vertices for BP scheduling.
+    val colorG = colorGraph(g)
+    val numColors: Int = colorG.vertices.select("color").distinct.count().toInt
+
+    // Convert GraphFrame to GraphX, and initialize beliefs.
+    val gx0 = colorG.toGraphX
+    // Schema maps for extracting attributes
+    val vColsMap = colorG.vColsMap
+    val eColsMap = colorG.eColsMap
+    // Convert vertex attributes to nice case classes.
+    val gx1: Graph[VertexAttr, Row] = gx0.mapVertices { case (_, attr) =>
+      // Initialize belief at 0.0
+      VertexAttr(attr.getDouble(vColsMap("a")), 0.0, attr.getInt(vColsMap("color")))
```

Comment:
one comment I forgot: I forgot if in the variational approximation, the coeffs should be random at start, or fixed at 0